### PR TITLE
feat(local-evidence): record local validation evidence during an Actions outage

### LINKED ISSUE #2323: feat(local-evidence): record local validation evidence during an Actions outage
## Background

All four required status checks on this repository's `main` ruleset
(`lint`, `idd-doctor`, `pnpm-boundary`, `idd-advisory-convergence`) are
GitHub Actions workflows, so an Actions outage makes every pull request
unmergeable at the platform level. No repository-level evidence can
change that: branch protection is enforced by GitHub, not by IDD, and
any design implying otherwise would be false.

What IDD can do is keep the work *verified* while the queue waits, so
recovery is a rerun rather than a re-review, and so nothing merges
later on evidence that no longer describes the code. The repository
already runs the same substance locally through its configured
`pre-push-validate` command; today that run leaves no durable,
HEAD-pinned trace anyone can check afterwards.

Blocked by #2320

## Proposed change

Record local validation as first-class, HEAD-pinned, expiring evidence,
and state the merge posture honestly.

- Define a local validation evidence marker recording the exact command
  set that ran, the HEAD SHA it ran against, the authoring actor, the
  timestamp, and the outcome.
- Accept the evidence in IDD's own pre-merge reasoning only while an
  outage relief declaration for the `ci-actions` service is active, and
  only for checks whose local equivalent actually ran. Evidence covering
  a subset of the required checks relieves nothing.
- Never let the evidence report a required platform check as passing.
  Merge-readiness output continues to list every unavailable required
  check as a blocker; the evidence changes what is *known*, not what is
  *green*.
- State the resulting posture explicitly in the documentation: with
  required checks unavailable, merges queue. Restoring the platform
  rollup is an out-of-band privileged operation that stays outside the
  autonomous loop and outside this issue's scope.
- On recovery, drive the re-verification sweep from the evidence
  marker's HEAD SHA, so a queued pull request whose HEAD moved is
  re-validated rather than merged on stale evidence.

## Acceptance criteria

- [ ] A JSON Schema and helper resolve the evidence marker:
      HEAD-pinned, actor-trust-filtered, and expiry-bounded, with the
      contract registered in `docs/idd-helper-scripts.md`.
- [ ] Unit tests from offline fixtures assert no relief for evidence
      bound to a different HEAD, from an untrusted actor, past its
      expiry, covering only part of the required command set, or with
      no active `ci-actions` relief declaration — each with a
      machine-readable reason.
- [ ] A test asserts that merge-readiness output for a pull request
      carrying valid local evidence still lists every unavailable
      required check as a blocker, and that no code path derives a
      merge from the evidence.
- [ ] A test asserts the recovery sweep re-validates, rather than
      clears, a pull request whose HEAD advanced past the evidence.
- [ ] The documentation states the queueing posture and that the
      out-of-band privileged merge is not part of the autonomous loop.
- [ ] Behavior detail lands in `docs/`, the phase-instruction delta
      stays minimal, and `node scripts/audit-docs.mjs --check` passes
      the documentation byte budget.
- [ ] `pnpm run build` regenerates committed helper output, and
      `node scripts/validate-schemas.mjs` plus the repository's
      `pre-push-validate` command pass.

## Candidate files

- `src/scripts/local-validation-evidence.mts`
- `schemas/local-validation-evidence.schema.json`
- `src/scripts/pre-merge-readiness.mts`
- `src/scripts/policy-helpers.mts`
- `schemas/policy.schema.json`
- `schemas/pre-merge-readiness.schema.json`
- `.github/instructions/idd-pre-merge.instructions.md`
- `idd-template/.github/instructions/idd-pre-merge.instructions.md`
- `docs/idd-helper-scripts.md`
- `docs/policy-constants.md`
- `scripts/local-validation-evidence.mjs`
- `tests/`

---

_Autopilot suitability: 3 / 5 -- higher is more autopilot-suitable;
below the configured floor is human-oriented._

_Effort: M -- author-estimated size; a soft autopilot
selection tie-breaker only._

### PR DESCRIPTION
**Pull request by kurone-kito**
## Summary

Closes #2323: record local `pre-push-validate` output as first-class, HEAD-pinned, expiring evidence during a #2320-declared Actions outage, so a queue recovers on a rerun rather than a re-review.

- **New marker** `idd-local-validation-evidence` (HEAD-pinned, actor-trust-filtered, expiry-bounded via `localValidationEvidence.maxAge`, default `PT4H`) plus `local-validation-evidence.mts`'s `resolveLocalValidationEvidence` — reports evidence as `present` only while a #2320 provider-outage declaration for service `ci-actions` is active, evidence covers every named required check, and it hasn't expired.
- **Wired into `pre-merge-readiness.mts`** as a purely additive, caller-precomputed `localValidationEvidence` field on `buildPreMergeReadinessSummary` — mirroring the existing `copilotUnavailable`/`advisoryConvergenceHeadCommittedAt` caller-precomputed-option pattern. `computePreMergeReadinessBlockers` has no reference to the new field at all, so it structurally cannot remove or downgrade a required-check blocker.
- **Recovery**: `evaluateLocalValidationEvidenceRecovery` re-validates (never merges) a pull request whose HEAD advanced past the recorded evidence.
- **Docs**: new helper section in `docs/idd-helper-scripts.md`, a `Local Validation Evidence Defaults` table in `docs/policy-constants.md`, and a short queueing-posture note in `idd-pre-merge.instructions.md` (all mirrored to `idd-template/`).

**Deliberately out of scope** (see code comment in `pre-merge-readiness.mts`): wiring `collectPreMergeReadiness`'s live CLI orchestration to call `resolveLocalValidationEvidence` automatically. That requires the already-resolved `ci.requiredCheckNames` (only known *after* `buildPreMergeReadinessSummary` runs) plus a second network fetch for the outage-declaration-target issue's comments — a circular-dependency/wiring concern best left to a follow-up, matching how #2320 itself shipped `evaluateProviderOutageRelief` as a standalone primitive without wiring it into `pre-merge-readiness.mts`'s live orchestration either. The `local-validation-evidence.mjs --resolve` CLI is fully functional standalone today.

## Test plan

- [x] `pnpm test` — 4150/4150 pass, including new `tests/local-validation-evidence.test.mts` (22 tests: round-trip, wrong-HEAD, untrusted actor, expiry boundary, partial coverage, failing outcome, no active declaration, malformed marker, recovery) and a dedicated `tests/pre-merge-readiness.test.mts` test asserting evidence is additive-only and never clears a required-check blocker.
- [x] `pnpm run typecheck`, `node scripts/validate-schemas.mjs`, `node scripts/audit-docs.mjs --check`, `node scripts/audit-code-span-wrap.mjs`, `pnpm run build:check`, `node scripts/idd-doctor.mjs` all clean.
- [x] `npx biome check`, `npx dprint check`, `npx markdownlint-cli2`, `npx cspell lint` all clean.

Refs #2320.

## Summary by CodeRabbit

* **New Features**
  * Added a command to record and resolve local validation evidence.
  * Evidence checks include commit alignment, freshness, coverage, validity, and trusted authorship.
  * Added informational evidence to pre-merge readiness without changing required-check blockers.
  * Added configurable evidence expiration, defaulting to four hours.
  * Added an optional secondary quiet-window gate before settled-review treatment.

* **Documentation**
  * Documented usage, evidence rules, configuration, schemas, recovery, and non-gating limitations.

* **Tests**
  * Added coverage for parsing, validation, expiration, recovery, CLI modes, schemas, and readiness reporting.

### CODE REVIEW

**Comment by kurone-kito** [comment]
_claude-idd2-orch: review triage snapshot — IDD automation marker. Do not edit._

**Comment by kurone-kito** [comment]
**Rejected** — chatgpt-codex-connector[bot] did not review HEAD a365ccdbeaf4e3219639693c0d4a8a4103c7cdfb (Codex usage limits for code reviews reached); this is not a completed review (source: #issuecomment-5491106628)

**Comment by kurone-kito** [comment]
**Accepted** — coderabbitai[bot] summary walkthrough at HEAD a365ccdbeaf4e3219639693c0d4a8a4103c7cdfb; actionable comments, if any, are dispositioned as their own review threads

File: src/scripts/local-validation-evidence.mts [hunk 1]
```diff
@@ -0,0 +1,647 @@
...
+  if (entries.length === 0) return null;
+  return [...entries].sort((a, b) => a.createdAt.localeCompare(b.createdAt))[
+    entries.length - 1
+  ];
+}
+
+function coversAll(covers: string[], requiredCheckNames: string[]): boolean {
+  const coveredSet = new Set(covers.map((entry) => entry.trim()));
+  return requiredCheckNames.every((name) => coveredSet.has(name.trim()));
+}
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/local-validation-evidence.mts:104]: **Accepted** — `coversAll()` returned true vacuously when `requiredCheckNames` is empty (e.g. the CLI resolve default when `--required-checks` is omitted), letting `present: true` report even though nothing was actually checked. Fixed: an empty `requiredCheckNames` now always returns `false` from `coversAll`. Test added: "an empty requiredCheckNames list never vacuously reports present".

File: src/scripts/protocol-helpers.mts [hunk 2]
```diff
@@ -6530,6 +6530,15 @@ export function buildPreMergeReadinessSummary(
...
     // `collectPreMergeReadiness` always sources the policy value.
     staleAgeMs?: number;
+    // #2323: the caller-precomputed `resolveLocalValidationEvidence`
+    // result (local-validation-evidence.mts), reported verbatim as its own
+    // top-level field -- purely informational. `computePreMergeReadinessBlockers`
+    // never reads this field, so it can never derive a blocker (or remove
+    // one) from local evidence; the required checks it references stay
+    // exactly as unresolved/unavailable as `ci` independently computed them.
+    // Omitted by unit callers and every caller that predates this option
+    // (field stays `null`, unchanged pre-#2323 behavior).
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/protocol-helpers.mts]: **Accepted** — the doc comment said the field "stays `null`" but the implementation actually omits it entirely when not supplied. Fixed the comment to say "omitted entirely -- not even `null`".

File: src/scripts/marker-helpers.mts [hunk 3]
```diff
@@ -1026,6 +1051,80 @@ export function parseProviderOutageAdvancedComment(
...
+    .map((entry) => String(entry ?? '').trim())
+    .filter(Boolean);
+  const outcome = String(payload?.outcome ?? '').trim();
+  if (
+    !actor ||
+    !/^[0-9a-f]{40}$/.test(headSha) ||
+    !commandSet ||
+    coversList.length === 0 ||
+    (outcome !== 'pass' && outcome !== 'fail')
+  ) {
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/marker-helpers.mts:1082]: **Accepted** — confirmed: joining raw covers entries with `,` before encoding meant a literal comma inside one entry's own name became indistinguishable from the join separator on decode, causing mis-split entries. Fixed: each covers entry is now percent-encoded individually before joining with a raw `,`, and the parser splits on the raw `,` before decoding each piece. Round-trip test added for a comma-containing check name.

File: src/scripts/marker-helpers.mts [hunk 4]
```diff
@@ -1026,6 +1051,80 @@ export function parseProviderOutageAdvancedComment(
...
+    throw new Error('invalid local validation evidence payload');
+  }
+  const encodedCommandSet = encodeExternalCheckWaiverField(commandSet);
+  const encodedCovers = encodeExternalCheckWaiverField(coversList.join(','));
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/marker-helpers.mts]: **Accepted** — same finding as Copilot's on this file (comma-in-entry ambiguity). Fixed: each covers entry is now percent-encoded individually before joining with a raw `,`; the parser splits on the raw `,` first, then decodes each piece. Round-trip test added for a comma-containing check name (tests/local-validation-evidence.test.mts, not schema-output-roundtrip.test.mts as suggested -- that file already covers the marker-shape round-trip, this behavioral edge case belongs with the marker's own test suite).

File: src/scripts/local-validation-evidence.mts [hunk 5]
```diff
@@ -0,0 +1,647 @@
...
+      trustedMarkerLogins,
+      outageDeclarationActive: outageDeclaration.active,
+      policy,
+      now,
+    });
+    render(result, args.format);
+    return { exitCode: 0, result };
+  }
+
+  // --record mutates; requires --apply.
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/local-validation-evidence.mts]: **Accepted** — clarified: without `--apply` this is a dry-run that only prints the rendered marker body; `--apply` is required only to post it to GitHub.

File: src/scripts/local-validation-evidence.mts [hunk 6]
```diff
@@ -0,0 +1,655 @@
...
+    reason = `no local validation evidence found for HEAD ${prHeadSha || '(unknown)'}`;
+  } else if (outcomeFail.length > 0) {
+    reason = 'latest local validation evidence recorded a failing outcome';
+  } else if (partialCoverage.length > 0) {
+    const missing = requiredCheckNames.filter(
+      (name) =>
+        !new Set(partialCoverage[partialCoverage.length - 1].covers).has(name),
+    );
+    reason = `local validation evidence covers only a subset of the required checks (missing: ${missing.join(', ') || 'unknown'})`;
+  } else if (expired.length > 0) {
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/local-validation-evidence.mts:243]: **Accepted** — when `requiredCheckNames` is empty, the reason now explicitly says "no required check names were given to resolve coverage against (pass --required-checks)" instead of the misleading "missing: unknown".

File: src/scripts/local-validation-evidence.mts [hunk 7]
```diff
@@ -0,0 +1,655 @@
...
+    return;
+  }
+  process.stdout.write(`${renderText(value)}\n`);
+}
+
+function printUsage(): void {
+  process.stdout.write(`usage: node scripts/local-validation-evidence.mjs --pr <number> --head-sha <40-hex> [options]
+
+Modes (default: resolve evidence validity for --pr/--head-sha):
+  --record                          post a new evidence marker (requires --apply)
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/local-validation-evidence.mts]: **Accepted** — missed this second occurrence in the previous round's fix. Reworded the --help usage line to "render (or, with --apply, post) a new evidence marker", matching the earlier fix to the code comment.

**Comment by kurone-kito** [comment]
**Accepted** — coderabbitai[bot] summary walkthrough at HEAD d9048f765f442a93ed0eb8ab3943fefb67571bae; actionable comments, if any, are dispositioned as their own review threads

File: src/scripts/local-validation-evidence.mts [hunk 8]
```diff
@@ -0,0 +1,662 @@
...
+    if (requiredCheckNames.length === 0) {
+      reason =
+        'no required check names were given to resolve coverage against (pass --required-checks)';
+    } else {
+      const missing = requiredCheckNames.filter(
+        (name) =>
+          !new Set(partialCoverage[partialCoverage.length - 1].covers).has(
+            name,
+          ),
+      );
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/local-validation-evidence.mts:240]: **Accepted** — confirmed: the missing-check diagnostic compared an untrimmed `name` against a trimmed covers Set, so a required-check name with surrounding whitespace would incorrectly report as missing even when actually covered. Fixed to trim both sides exactly like `coversAll` does. Test added.

File: src/scripts/local-validation-evidence.mts [hunk 9]
```diff
@@ -0,0 +1,662 @@
...
+  } else if (outcomeFail.length > 0) {
+    reason = 'latest local validation evidence recorded a failing outcome';
+  } else if (partialCoverage.length > 0) {
+    if (requiredCheckNames.length === 0) {
+      reason =
+        'no required check names were given to resolve coverage against (pass --required-checks)';
+    } else {
+      const missing = requiredCheckNames.filter(
+        (name) =>
+          !new Set(partialCoverage[partialCoverage.length - 1].covers).has(
```
[the remark this answers is not in the document]
Comment (kurone-kito) [reply src/scripts/local-validation-evidence.mts]: **Accepted** — same finding as Copilot's on this line. Fixed: the missing-check diagnostic now trims both sides exactly like `coversAll` does. Test added.

**Comment by kurone-kito** [comment]
**Accepted** — coderabbitai[bot] summary walkthrough at HEAD 1a3fa27200d092766853833cc280b2331f027a69; actionable comments, if any, are dispositioned as their own review threads

**Comment by kurone-kito** [comment]
_claude-idd2-orch: review triage snapshot — IDD automation marker. Do not edit._

**Comment by kurone-kito** [comment]
_claude-idd2-orch: review triage snapshot — IDD automation marker. Do not edit._
